### DISCUSSION TITLE
how to handle list of files on an ftp server as a target

### DISCUSSION THREAD

**Discussion by majazaloznik** [ropensci/targets#417] [Help]
This is a complete noob question: I want to check the list of files on an ftp server as my first target to see if there's something new there compared to the files I already have downloaded. How would I go about ensuring that target is always run?

**Comment by wlandau** [comment]
A target with `format = "file"` can track one or more input files or directories. Details are explained at https://books.ropensci.org/targets/files.html#external-input-files.

**Reply by majazaloznik** [reply]
thanks, but sorry, maybe i wasn't clear: 
i have a function `ftp_list_files()` that gets the file list on an ftp server 
i have a second function `get_new_files()` that compares the file list with what i have locally and downloads anything that is missing locally. 

it's the first function that i need to run every time i run `tar_make()`

**Reply by wlandau** [reply]
To run a target on every `tar_make()`, define it with `cue = tar_cue(mode = "always")` in `tar_target()`.

**Reply by majazaloznik** [reply]
That's perfect Will, exactly what I was after, cheers!

**Reply by majazaloznik** [reply]
hmm, actually no, i'm not getting what i'd like unfortunately..
now `ftp_list_files()` gets run every time, but if there's been no change, the target is still seen as updated. instead i want the whole pipeline to stop if there's been no change.

**Reply by wlandau** [reply]
In that case, you could create a global object for the FTP file list upstream of all the targets. The list would get refreshed continuously but not necessarily invalidate anything.

```r
# _targets.R
library(targets)
existing_files <- ftp_list_files()
list(
  tar_target(all_files, get_new_files(existing_files), format = "file"),
  tar_target(analysis, analyze_files(all_files)),
  ...
)
```

**Reply by majazaloznik** [reply]
haha, of course! i've just discovered this wonderful hammer that is `targets` so now everything looks like a nail to me :)
thanks for your help!

**Reply by majazaloznik** [reply]
ok, i've got the pipeline i wanted, but i'm somehow not happy with it conceptually. it feels like `targets` should be able to handle this logic, not that I solve it externally. the point is that if there are no new files on the server, the rest of the pipeline doesn't have to be run. 
```
library(targets)
newly_available_files <- ftp_list_new_files()
list(
  tar_target(new_files, get_new_files(newly_available_files), format = "file"),
  tar_target(analysis, analyze_files(new_files, old_files)), 
  ...
)
if(length(newly_available_files) != 0) tar_make()
```

**Reply by wlandau** [reply]
`targets` has trouble thinking about just the new files. When you first run the pipeline, there will be some new files, and all the targets will run. But if you run `tar_make()` and there are no new files, the `newly_available_files` will have changed, and `tar_make()` will decide to run the downstream targets.

I recommend keeping track of the existing files, including the new and old files, whatever those happen to be. That way, when new files arrive, `targets` will notice a change to `existing_files` and rerun the affected parts of the pipeline. Otherwise, if no new files arrived, `existing_files` will not change, and `tar_make()` will skip all the targets if there are no other sources of change.
